x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/api/perform_rule_upgrade/assert_upgradeable_fields_match_target_type.ts

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { DIFFABLE_RULE_TYPE_FIELDS_MAP } from '../../../../../../common/api/detection_engine';
+
+/**
+ * Validates that the upgradeable (diffable) fields match the target rule type's diffable fields.
+ *
+ * This function is used in the rule upgrade process to ensure that the fields
+ * specified for upgrade in the request body are valid for the target rule type.
+ * It checks each upgradeable field provided in body.rule[].fields against the
+ * set of diffable fields for the target rule type.
+ *
+ * @param {string[]} upgradeableFields - An array of field names to be upgraded.
+ * @param {string} targetRuleType - The target rule type (e.g., 'query', 'eql', 'machine_learning').
+ * @throws {Error} If an upgradeable field is not valid for the target rule type.
+ *
+ * @examples
+ * assertUpgradeableFieldsMatchTargetType(['kql_query', 'severity'], 'query');
+ * assertUpgradeableFieldsMatchTargetType(['esql_query', 'description'], 'esql');
+ * assertUpgradeableFieldsMatchTargetType(['machine_learning_job_id'], 'eql'); // throws error
+ *
+ * @see {@link DIFFABLE_RULE_TYPE_FIELDS_MAP} in diffable_rule.ts for the mapping of rule types to their diffable fields.
+ */
+export const assertUpgradeableFieldsMatchTargetType = (
```

Comment:
I get your point, but it's not exactly so, since this function checks that the **diffable** or **upgradeable** fields match the target type, not the rules belonging to the plain schema (so `data_source` rather than `index`, `eql_query` rather than `query`, etc.)

I think the name still needs to distinguish this difference. What do you think of something like:
```ts
export const assertDiffableFieldsMatchTargetType = (
  diffableFields: string[],
  targetRuleType: string
) => {
  //etc
```
?

Reply:
I applied that change, let me know if you think it's not enough